src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/InstrumentRecorder.cs

```diff
@@ -0,0 +1,200 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Metrics;
+using System.Threading;
+
+namespace System.Diagnostics.Metrics
+{
+    /// <summary>
+    /// A helper class to record the measurements published from <see cref="Instrument{T}"/> or <see cref="ObservableInstrument{T}"/>.
+    /// </summary>
+    public sealed class InstrumentRecorder<T> : IDisposable where T : struct
+    {
+        private bool _isObservableInstrument;
+        private bool _disposed;
+        private Instrument? _instrument;
+
+        private readonly MeterListener _meterListener;
+        private readonly List<Measurement<T>> _measurements;
+
+        /// <summary>
+        /// Initialize a new instance <see cref="InstrumentRecorder{T}" /> to record the measurements published by <see cref="Instrument" />.
+        /// </summary>
+        /// <param name="instrument">The <see cref="Instrument" /> to record measurements from.</param>
+        public InstrumentRecorder(Instrument instrument)
+        {
+            if (instrument is null)
+            {
+                throw new ArgumentNullException(nameof(instrument));
+            }
+
+            if (instrument is not Instrument<T> and not ObservableInstrument<T>)
+            {
+                throw new InvalidOperationException(SR.InvalidInstrumentType);
+            }
+
+            _measurements = new List<Measurement<T>>();
+
+            _meterListener = new MeterListener();
+            _meterListener.InstrumentPublished = (inst, listener) =>
```

Comment:
I think you should only need to do this, listening for a publish notification should be unnecessary.
```suggestion
            _meterListener.SetMeasurementEventCallback<T>(OnMeasurementRecorded);
            _meterListener.EnableMeasurementEvents(instrument, state: null);
```

Reply:
Thanks, I have done that.